<p>Genes can be converted (<a href="https://en.wikipedia.org/wiki/Translation_(biology)" target="_blank" rel="noopener">translated</a>) to proteins using&nbsp;a three base <a href="https://en.wikipedia.org/wiki/Genetic_code" target="_blank" rel="noopener">decoding system</a> as described in <a href="../../../255/" target="_blank" rel="noopener">Bite 255</a>.</p>
<p>Your job is to create a function that takes a <a href="https://en.wikipedia.org/wiki/Coding_region" target="_blank" rel="noopener">coding sequence (CDS)</a>&nbsp;(=the region of a gene that encodes a protein but excluding any other features) and returns the translated&nbsp;protein/polypeptide as a <code>str</code>.</p>
<p>For your convenience, we've installed <a href="https://biopython.org/" target="_blank" rel="noopener">Biopython</a> for you. Have a look at the <a href="https://biopython.org/docs/1.75/api/Bio.Seq.html" target="_blank" rel="noopener">Bio.Seq Module</a> and locate the correct function to&nbsp;solve this bite. You will also need to ensure proper cleaning up of input whitespace.&nbsp;</p>
<p>Although the genetic code is very conserved, there are differences between encodings in certain organisms and cell compartments. Therefore, there are different mappings or "<a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi" target="_blank" rel="noopener">translation tables</a>" implemented in Biopython that you can directly use.</p>
<p>Also make sure you find and set the option to use&nbsp;complete coding sequences (CDS) as this means that start codons in position 1 are evaluated as the amino&nbsp;acid M instead of their usual counterpart.</p>
<p>Example of how the function should work:</p>
<pre>&gt;&gt;&gt; translate_cds("ATGGGGTTTTAA", "Bacterial")<br />'MGF'<br />&gt;&gt;&gt; translate_cds("TTGGGGTTTTAA", "Bacterial")<br />'MGF'<br />&gt;&gt;&gt; translate_cds("ACGGGGTTTTAA", "Bacterial")<br />[...]<br />TranslationError: First codon 'CTG' is not a start codon</pre>
<p>Good luck!</p>
<p>&nbsp;</p>